mypyc/irbuild/function.py

```diff
@@ -833,60 +850,58 @@ def generate_singledispatch_dispatch_function(
     current_func_decl = builder.mapper.func_to_decl[fitem]
     arg_info = get_args(builder, current_func_decl.sig.args, line)
 
-    def gen_func_call_and_return(
-        func_name: str,
-        fdef: FuncDef,
-        fullname: Optional[str] = None
-    ) -> None:
-        if is_decorated(builder, fdef):
-            func = load_func(builder, func_name, fullname, line)
-            ret_val = builder.builder.py_call(
-                func, arg_info.args, line, arg_info.arg_kinds, arg_info.arg_names
-            )
-        else:
-            func_decl = builder.mapper.func_to_decl[fdef]
-            ret_val = builder.builder.call(
-                func_decl, arg_info.args, arg_info.arg_kinds, arg_info.arg_names, line
-            )
+    def gen_native_func_call_and_return(fdef: FuncDef) -> None:
+        func_decl = builder.mapper.func_to_decl[fdef]
+        ret_val = builder.builder.call(
+            func_decl, arg_info.args, arg_info.arg_kinds, arg_info.arg_names, line
+        )
         coerced = builder.coerce(ret_val, current_func_decl.sig.ret_type, line)
         builder.nonlocal_control[-1].gen_return(builder, coerced, line)
 
-    # Add all necessary imports of other modules that have registered functions in other modules
-    # We're doing this in a separate pass over the implementations because that avoids the
-    # complexity and code size implications of generating this import before every call to a
-    # registered implementation that might need this imported
-    for _, impl in impls:
-        if not is_decorated(builder, impl):
-            continue
-        module_name = impl.fullname.rsplit('.')[0]
-        if module_name not in builder.imports:
-            # We need to generate an import here because the module needs to be imported before we
-            # try loading the function from it
-            builder.gen_import(module_name, line)
-
-    # Sort the list of implementations so that we check any subclasses before we check the classes
-    # they inherit from, to better match singledispatch's behavior of going through the argument's
-    # MRO, and using the first implementation it finds
-    for dispatch_type, impl in sort_with_subclasses_first(impls):
+    registry = load_singledispatch_registry(builder, fitem, line)
+    find_impl = builder.load_module_attr_by_fullname('functools._find_impl', line)
```

Comment:
The core logic here looks really solid, but we definitely definitely need to add a cache in front of `_find_impl` (like the real `singledispatch` does), which looks like it is wildly slow (mostly because of various stuff it does to handle abstract base classes -- though even just the pycall we have to make to it is probably enough for me to want a cache here).

We want the happy case to be:
 1. Look up `arg_type` in our cache dictionary
 2. It is present and an integer, great
 3. Dispatch and do a native call

The slightly less happy case is it not being an int, and needing to do a pycall.
The sad case is:
 1. Look up `arg_type` in our cache dictionary
 2. Oh no, it isn't there!
 3. Call `_find_impl`
 4. Put the result into our cache dictionary
 5. Do the call stuff